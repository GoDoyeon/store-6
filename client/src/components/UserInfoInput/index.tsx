import React, { useState, useEffect } from 'react';
import {
  StyledUserInfoInput,
  UserInfoInputTitle,
  ModifyButton,
} from './index.style';

const DEFAULT_PLACEHOLDER = '수정은 수정 버튼으로';

interface Props {
  title: string;
  value?: string;
  defaultDisabled?: boolean;
  placeholder?: string;
  validator?: (userInput: string) => boolean;
  showWarning?: () => void;
  onSubmit?: (userInput: string) => void;
  inputComponent: React.ReactNode;
}

const UserInfoInput: React.FC<Props> = ({
  title,
  value,
  defaultDisabled,
  placeholder,
  validator,
  showWarning,
  onSubmit,
  inputComponent,
}) => {
  const safeValidator = (valueToValidate: string) =>
    validator ? validator(valueToValidate ?? '') : true;

  const [userInput, setUserInput] = useState(value ?? '');
  const [isCanSubmit, setIsCanSubmit] = useState(safeValidator(value));
  const [disabled, setDisabled] = useState(defaultDisabled ?? true);

  const handleOnChangeUserInput = (inputValue: string) => {
    setIsCanSubmit(safeValidator(inputValue));
    setUserInput(inputValue);
  };

  const handleOnClickModifyButton = () => {
    if (disabled) {
      setDisabled(false);
      return;
    }
    if ((validator && !validator(userInput)) || !isCanSubmit) {
      showWarning();
      return;
    }
    if (onSubmit) {
      onSubmit(userInput);
      setDisabled(true);
      setIsCanSubmit(safeValidator(value));
    }
  };

  useEffect(() => {
    setIsCanSubmit(safeValidator(value));
    setUserInput(value);
  }, [value]);

  return (
    <StyledUserInfoInput>
      <UserInfoInputTitle>{title}</UserInfoInputTitle>
      {typeof inputComponent === 'function' &&
        inputComponent({
          disabled,
          value: disabled ? '' : userInput,
          onChange: handleOnChangeUserInput,
          placeholder: placeholder?.length ? placeholder : DEFAULT_PLACEHOLDER,
        })}
      <ModifyButton
        disabled={disabled || !isCanSubmit}
        onClick={handleOnClickModifyButton}
      >
        {disabled ? '수정' : '완료'}
      </ModifyButton>
    </StyledUserInfoInput>
  );
};

export default UserInfoInput;
