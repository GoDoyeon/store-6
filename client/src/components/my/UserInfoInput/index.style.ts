import styled from 'styled-components';

export const StyledUserInfoInput = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-evenly;
  width: 630px;
  height: 70px;
  margin: 30px 0;
`;

export const UserInfoInputTitle = styled.div`
  font-size: 20px;
`;

export const ModifyButton = styled.div<{
  disabled: boolean;
}>`
  border: none;
  outline: none;
  font-size: 15px;
  cursor: pointer;
  color: ${({ disabled }) => (disabled ? '#000' : '#2ac1bc')};
`;
