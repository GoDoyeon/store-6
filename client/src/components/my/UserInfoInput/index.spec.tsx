import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import UserInfoInput from './index';

describe('<UserInfoInput />', () => {
  it('should render same with snapshot', () => {
    const { container } = render(
      <UserInfoInput title="test" inputComponent={() => {}} />,
    );
    expect(container).toMatchSnapshot();
  });
});
