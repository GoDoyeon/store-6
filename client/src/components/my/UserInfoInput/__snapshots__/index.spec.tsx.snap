// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserInfoInput /> should render same with snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  width: 630px;
  height: 70px;
  margin: 30px 0;
}

.c1 {
  font-size: 20px;
}

.c2 {
  border: none;
  outline: none;
  font-size: 15px;
  cursor: pointer;
  color: #000;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      test
    </div>
    <div
      class="c2"
      disabled=""
    >
      수정
    </div>
  </div>
</div>
`;
